### PR TITLE
Add dired-sidebar-one-instance

### PR DESCRIPTION
**Pull request by fasheng**
Only show one buffer instance for dired-sidebar for each frame, hope helps~

### CODE REVIEW

**Comment by jojojames** [comment]
Hmnn... I'm not sure about this one just yet.

I wonder what happens if two frames share the same sidebar and then frame 2 opens another sidebar. Wouldn't the first buffer go away in frame 1?

**Comment by fasheng** [comment]
If two frames share the same sidebar, just like `split-window` did for one buffer, they will not effect each other except you kill one or change directory.

And because you fix the #29, I update the code to use `find-alternate-file` instead which like my hack code did at first.

You could hava a try, one instance just works fine and switching buffers is more easily ;)

**Comment by jojojames** [comment]
Lemme give it a try first when I find some free time (no promises on when :)). 

Thanks for making these changes anyway! Great to have someone digging in and tweaking some code. :D

**Comment by jojojames** [comment]
I think you need to rebase, by the way. :)

**Comment by fasheng** [comment]
Oh, sorry. Rebased now, and works fine for my emacs 26.2. Hope could be merged this time :)

File: dired-sidebar.el [hunk 1]
```diff
@@ -322,6 +322,11 @@ For more information, look up `delete-other-windows'."
   :type 'boolean
   :group 'dired-sidebar)
 
+(defcustom dired-sidebar-one-instance nil
```
Comment (jojojames) [inline dired-sidebar.el]: This should be dired-sidebar-one-instance-p.

File: dired-sidebar.el [hunk 2]
```diff
@@ -553,12 +558,15 @@ With universal argument, use current directory."
   (interactive)
   (if (dired-sidebar-showing-sidebar-p)
       (dired-sidebar-hide-sidebar)
-    (let* ((file-to-show (dired-sidebar-get-file-to-show))
+    (let* ((old-buf (dired-sidebar-buffer (selected-frame)))
```
Comment (jojojames) [inline dired-sidebar.el]: Nit: Can you change this from old-buf to old-buffer?

File: dired-sidebar.el [hunk 3]
```diff
@@ -696,8 +704,10 @@ window selection."
                (progn
                  (switch-to-buffer buf-name)
                  (dired-sidebar-update-state (current-buffer)))
-             ;; Copied from `dired-find-file'.
-             (find-file dired-file-name)
+             (if (and dired-sidebar-one-instance (file-directory-p dired-file-name))
+                 (find-alternate-file dired-file-name)
```
Comment (jojojames) [inline dired-sidebar.el]: You're making use of the fact that find-alternate-file kills the dired-sidebar buffer right?

I think it might be more clear -maintenance wise- to use find-file together with kill-buffer.

File: dired-sidebar.el [hunk 4]
```diff
@@ -740,7 +750,9 @@ Select alternate window using `dired-sidebar-alternate-select-window-function'."
          (progn
            (switch-to-buffer up-name)
            (dired-sidebar-update-state (current-buffer)))
-       (dired-up-directory)
+         (if dired-sidebar-one-instance
+             (find-alternate-file "..")
```
Comment (jojojames) [inline dired-sidebar.el]: Same comment as above. This time it's about dired-up-directory vs using find-alternate-file.

Admittedly, I'm not too against it but this code can be hard to follow already since there's a lot of conditionals at this point.

File: dired-sidebar.el (hunk 1 above)
Comment (fasheng) [reply dired-sidebar.el]: Got it.

File: dired-sidebar.el (hunk 2 above)
Comment (fasheng) [reply dired-sidebar.el]: Got it

File: dired-sidebar.el (hunk 3 above)
Comment (fasheng) [reply dired-sidebar.el]: `find-alternate-file` is better, because it will remember the last cursor when switching directories.

File: dired-sidebar.el (hunk 4 above)
Comment (fasheng) [reply dired-sidebar.el]: Same reason as above.

> Admittedly, I'm not too against it but this code can be hard to follow already since there's a lot of conditionals at this point.

I'm not sure how to improve this, you could help me after merged.

**Comment by fasheng** [comment]
Please review again~

**Comment by jojojames** [comment]
Hmnn, looks OK with me for now.

Thanks!
